File: gateway/src/main/java/com/beat/gateway/shared/internal/GatewayConfigImportSelector.java
package com.beat.gateway.shared.internal;

import java.util.List;
import java.util.Map;

import org.springframework.context.annotation.DeferredImportSelector;
import org.springframework.core.type.AnnotationMetadata;

import com.beat.gateway.EnableGatewayConfig;
import com.beat.gateway.GatewayConfigGroup;

public class GatewayConfigImportSelector implements DeferredImportSelector {

	@Override
	public String[] selectImports(AnnotationMetadata importingClassMetadata) {
		return getValues(importingClassMetadata).stream()
			.map(group -> group.getConfigClass().getName())
			.toArray(String[]::new);
	}

	private List<GatewayConfigGroup> getValues(AnnotationMetadata metadata) {
		Map<String, Object> attributes = metadata.getAnnotationAttributes(EnableGatewayConfig.class.getName());
		if (attributes == null) {
			return List.of();
		}

		GatewayConfigGroup[] values = (GatewayConfigGroup[])attributes.get("value");
		return values == null ? List.of() : List.of(values);
	}
}

File: support/security-web/src/main/kotlin/com/beat/support/security/shared/internal/GatewayConfigImportSelector.kt
package com.beat.support.security.shared.internal

import com.beat.support.security.EnableGatewayConfig
import com.beat.support.security.GatewayConfigGroup
import org.springframework.context.annotation.DeferredImportSelector
import org.springframework.core.type.AnnotationMetadata

internal class GatewayConfigImportSelector : DeferredImportSelector {

    override fun selectImports(importingClassMetadata: AnnotationMetadata): Array<String> =
        getValues(importingClassMetadata).map { group -> group.configClass.name }.toTypedArray()

    private fun getValues(metadata: AnnotationMetadata): List<GatewayConfigGroup> {
        val attributes =
            metadata.getAnnotationAttributes(EnableGatewayConfig::class.java.name)
                ?: return emptyList()

        val values = attributes["value"] as? Array<*> ?: return emptyList()
        return values.filterIsInstance<GatewayConfigGroup>()
    }
}
